tokio/src/coop.rs

```diff
@@ -147,9 +128,45 @@ cfg_rt! {
     }
 }
 
-cfg_coop! {
-    use std::task::{Context, Poll};
+pin_project! {
+    /// Future for the [`unconstrained`](unconstrained) method.
+    #[must_use = "Unconstrained does nothing unless polled"]
+    pub struct Unconstrained<F> {
+        #[pin]
+        inner: F,
+    }
+}
+
+impl<F> Future for Unconstrained<F>
+where
+    F: Future,
+{
+    type Output = <F as Future>::Output;
 
+    fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+        let inner = self.project().inner;
+        with_budget(Budget::unconstrained(), || inner.poll(cx))
+    }
+}
+
+/// Turn off cooperative scheduling for a future. The future or stream will never yield.
+///
+/// # Examples
+///
+/// ```
+/// # #[tokio::main]
+/// # async fn main() {
+/// use tokio::coop;
+///
+/// let fut = async { () };
+/// coop::unconstrained(fut).await;
+/// # }
+/// ```
+pub fn unconstrained<F>(inner: F) -> Unconstrained<F> {
```

Comment:
Should this only be available under `cfg_coop!` @carllerche ?

Reply:
Wouldn't that mean that if you need to use this in a library you'd force your users to enable coop? (granted, I'm not sure familiar with features in Cargo, as I mentioned in that other thread we don't really use Cargo at $work :) )

Reply:
Ah, I remember now. [`cfg_coop!`](https://github.com/tokio-rs/tokio/blob/8efa62013b551d5130791c3a79ce8ab5cb0b5abf/tokio/src/macros/cfg.rs#L346) is actually a short-hand for "if any of these features are enabled". It's not a feature in and of itself. The idea is that if you don't bring in a runtime or any resource types, then there's no reason to build the coop stuff. So I think `cfg_coop!` should also apply to these types.

Reply:
That's a relatively complicated feature list. I might just prefer always exposing it, having it be a no-op without coop in the runtime.

Reply:
I think this was at the request of @carllerche, though his opinions may have changed since?

Reply:
My gut is that this should live in the task module for the public API. I like what @Darksonn suggests, expose the fn anytime the task module is enabled but the impl changes based on cfg_coop. Either it does something or it is a noop.

Reply:
> My gut is that this should live in the task module for the public API. 

Should I be making this update here? 

> the impl changes based on cfg_coop

Reply:
> My gut is that this should live in the task module for the public API. I like what @Darksonn suggests, expose the fn anytime the task module is enabled but the impl changes based on cfg_coop. Either it does something or it is a noop.

Should I be making this update in this PR then? I'm not certain exactly what y'all would like me to change to get this moving forward :)

Reply:
Yes, please go ahead and make this change.

Reply:
Sounds good; thanks; I just updated the PR.